### PR TITLE
Chunkify

### PR DESCRIPTION
**Pull request by tristpinsm**

### CODE REVIEW

**Comment by tristpinsm** [comment]
One thing I just noticed : In `Truncate`, where it uses the weights to determine the precision, at the moment it makes a copy of the weights array to scale by the fractional variance limit. We could avoid this copy by adding the limit factor as an argument to the `bit_truncate_weight` function. Is that worth adding to this PR?

**Comment by jrs65** [comment]
> One thing I just noticed : In `Truncate`, where it uses the weights to determine the precision, at the moment it makes a copy of the weights array to scale by the fractional variance limit. We could avoid this copy by adding the limit factor as an argument to the `bit_truncate_weight` function. Is that worth adding to this PR?

That might be worth adding if you are making other changes. But I wouldn't bother otherwise.

**Comment by jrs65** [comment]
@tristpinsm I'm wondering what the status of this branch is. I guess maybe we should wait until @nritsche has finished his new zarr compression support?

**Comment by nritsche** [comment]
> @tristpinsm I'm wondering what the status of this branch is. I guess maybe we should wait until @nritsche has finished his new zarr compression support?

If not specified, zarr will guess the chunk sizes, so I'm not sure if we still need this.
(https://zarr.readthedocs.io/en/stable/tutorial.html#chunk-optimizations)

**Comment by jrs65** [comment]
I think we may always want to set the chunk size. Setting it appropriately is key to achieving a good compression ratio.

**Comment by tristpinsm** [comment]
I haven't touched this since then. I guess it would need to be integrated with the changes to the truncation routines now.

File: draco/core/containers.py [hunk 1]
```diff
@@ -265,6 +265,32 @@ def add_dataset(self, name):
...
+
+        For every dataset, check if chunks and compression are set, and
+        if not set them to dataset_spec values.
+        """
+        for dset in self.dataset_spec:
+            if "chunks" in self.dataset_spec[dset] and self[dset].chunks is None:
+                # ensure chunks aren't larger than dataset shape
+                chunks = ()
+                for i, l in enumerate(self[dset].shape):
+                    chunks += (min(self.dataset_spec[dset]["chunks"][i], l),)
```
Comment (nritsche) [inline draco/core/containers.py:279]: What is this doing?
Why would this be below 1 or why would you choose chunk size 1?

Comment (tristpinsm) [reply draco/core/containers.py:279]: did you misread `l` as `1`?

Comment (nritsche) [reply draco/core/containers.py:279]: Thanks

**Comment by nritsche** [comment]
I'm re-using this in #130

**Comment by tristpinsm** [comment]
> I'm re-using this in #130

hooray, Chunkify lives!

**Comment by jrs65** [comment]
@nritsche shall we close this branch and wrap it up into yours?

**Comment by jrs65** [comment]
> @nritsche shall we close this branch and wrap it up into yours?

Ping!

**Comment by nritsche** [comment]
> @nritsche shall we close this branch and wrap it up into yours?

I was planning to close this once mine is merged, but feel free to do it now.

**Comment by jrs65** [comment]
This is subsumed into Rick's zarr branch that needs to be merged.

**Pull request closed without merging**